src/main/java/foi/air/szokpt/accountmng/services/UserService.java

```diff
@@ -0,0 +1,43 @@
+package foi.air.szokpt.accountmng.services;
+
+import foi.air.szokpt.accountmng.entitites.User;
+import foi.air.szokpt.accountmng.entitites.UserRole;
+import foi.air.szokpt.accountmng.exceptions.ValidationException;
+import foi.air.szokpt.accountmng.repositories.RoleRepository;
+import foi.air.szokpt.accountmng.repositories.UserRepository;
+import foi.air.szokpt.accountmng.util.hashing.Hasher;
+import foi.air.szokpt.accountmng.util.validation.Validator;
+import org.springframework.stereotype.Service;
+
+import java.util.Optional;
+
+@Service
+public class UserService {
+    private final UserRepository userRepository;
+    private final RoleRepository roleRepository;
+    private final Validator<User> userDataValidator;
+    private final Hasher hasher;
+
+    public UserService(UserRepository userRepository, RoleRepository roleRepository, Validator<User> validator, Hasher hasher) {
+        this.userRepository = userRepository;
+        this.roleRepository = roleRepository;
+        this.userDataValidator = validator;
+        this.hasher = hasher;
+    }
+
+    public void registerUser(User user) {
+        user.setEmail(user.getEmail().toLowerCase());
+        userDataValidator.validateData(user);
+        assignRoleToUser(user);
+        user.setPassword(hasher.hashText(user.getPassword()));
+        userRepository.save(user);
```

Comment:
Tu bi trebao biti try catch implementirani

```suggestion
        try {
               userRepository.save(user);
        } catch (Exception e) {
               throw new ValidationException("Failed to register user.");
         }
```